File: ts-sdk/src/blocks/get.latest.block.number.ts
import { JsonRpcProvider } from "ethers";
import { Logger, WithRetry, assertExists } from "../utils";
import { Cache } from "../cache";

// returns the latest block number as reported by the "eth_blockNumber" json-rpc method
export type GetLatestBlockNumber = (
  chainId: number,
  provider: JsonRpcProvider
) => Promise<number>;

export function provideGetLatestBlockNumber(
  cache: Cache,
  withRetry: WithRetry,
  logger: Logger
): GetLatestBlockNumber {
  assertExists(cache, "cache");
  assertExists(withRetry, "withRetry");

  return async function getLatestBlockNumber(
    chainId: number,
    provider: JsonRpcProvider
  ) {
    // check cache first
    logger.debug(`checking cache for eth_blockNumber for chain ${chainId}...`);
    let cachedBlockNumberHex = await cache.getLatestBlockNumber(chainId);
    if (cachedBlockNumberHex) {
      logger.debug(
        `chain ${chainId} latest cached eth_blockNumber: ${parseInt(
          cachedBlockNumberHex
        )}`
      );
      return parseInt(cachedBlockNumberHex);
    }

    logger.debug(
      `falling back to bot's provider for eth_blockNumber for chain ${chainId}...`
    );
    // try fetching from given provider
    const blockNumberHex: string = await withRetry(
      provider.send.bind(provider),
      ["eth_blockNumber", []]
    );
    return parseInt(blockNumberHex);
  };
}

File: py-sdk/src/forta_bot_sdk/blocks/get_latest_block_number.py
from typing import Callable
from web3 import AsyncWeb3
from ..cache import Cache
from ..utils import WithRetry, Logger, assert_exists

GetLatestBlockNumber = Callable[[int, AsyncWeb3], int]


def provide_get_latest_block_number(
        cache: Cache,
        with_retry: WithRetry,
        logger: Logger):
    assert_exists(cache, "cache")
    assert_exists(with_retry, "with_retry")
    assert_exists(logger, "logger")

    async def get_latest_block_number(chain_id: int, provider: AsyncWeb3) -> int:
        # check cache first
        logger.debug(
            f'checking cache for eth_blockNumber for chain {chain_id}')
        cached_block_number_hex = await cache.get_latest_block_number(chain_id)
        if cached_block_number_hex:
            logger.debug(
                f'chain {chain_id} latest cached eth_blockNumber: {int(cached_block_number_hex, 0)}')
            return int(cached_block_number_hex, 0)

        logger.debug(
            f'falling back to bots provider for eth_blockNumber for chain {chain_id}')
        response = await with_retry(provider.provider.make_request, "eth_blockNumber", [])
        block_number = int(response['result'], 0)
        return block_number

    return get_latest_block_number
